Add jupyter integration for lets-plot-kotlin and lets-plot-kotlin-geotools

AndreiKingsley opened the pull request:
* Solves #244.

Add Jupyter Integration plot-api and toolkit/geotools modules. All the old logic (imports/init/renders) from descriptors has been moved into them. Updated descriptors are in [this PR](https://github.com/Kotlin/kotlin-jupyter-libraries/pull/421).

A new type returned by rendering has also been added, which allows you to use the following features that are now available for Kandy plots in Kotlin Notebook:
* Native rendering in Swing;
* Dynamic IDEA theme application for plots (only for plots where flavor is not specified);
* Export plot from cell in different formats (PNG, SVG, JPG, HTML);
* Copy plot (as PNG) from cell;

Added `letsPlotNotebookConfig` variable (in notebooks only)  which allows to disable native rendering in KTNB.

These changes do not affect Jupyter and Datalore notebooks.

Also this rendering adds static SVG in the output, hidden by script which allows to show static part of plot on GitHub/Gist.

Several simple tests for Jupyter Integration are also added.

AndreiKingsley commented:
I've run into the problem that the latest versions of the Kotlin jupyter api use java 11, which makes it completely incompatible with geotools module, and partially with plot-api. In particular, I had to add different versions of the plugin to them because of this. But it's clear that this is a bad solution. @ileasile suggested the following: https://github.com/Kotlin/kotlin-jupyter-libraries/pull/421#issuecomment-2399998252

ileasile commented:
I suggest adding lets-plot-jupyter and lets-plot-geotools-jupyter modules. They should have all dependencies that are now listed in corresponding descriptors, and also integration classes. Also, they should target 11th version of JVM bytecode. I'm ready to pair-program it with you if there are any problems in implementation

AndreiKingsley commented:
Shouldn't be difficult, but I'm interested in @alshan opinion.

alshan commented:
Jupyter integration doesn't belong to the "plot-api" project in my opinion. Let's move it to a separate project nested in the 'toolkit' folder? Can have java 11 as a target there.

alshan commented on plot-api/src/commonMain/kotlin/org/jetbrains/letsPlot/frontend/NotebookFrontendContext.kt:
```diff
@@ -50,6 +51,10 @@ class NotebookFrontendContext(
         return getDisplayHtml(plotBunch.toSpec())
     }
 
+    fun getHtml(figure: Figure): String {
```
Nice to have this method in 'Figure' instead.

AndreiKingsley replied:
But this method depends on `NotebookFrontendContext` and its behaviour is defined by `isolatedFrame`.  
May be it's better to add extensions `Figure.toHTML(): String`, `Figure.toSVG(): String`?

alshan replied:
Maybe we are talking about different use-cases. What is `NotebookFrontendContext.getHtml()` is intended for?

What I mean is an analog of [`to_html()`](https://lets-plot.org/python/pages/api/lets_plot.plot.core.PlotSpec.html#lets_plot.plot.core.PlotSpec.to_html) in Python where the output would likely have some external use and is configured via parameter `iframe` as needed.

AndreiKingsley replied:
Isn't that what `ggsave()` does (almost)? It seems logical that `NotebookFrontendContext` and its methods are responsible for the internal logic associated with the notebook . And with `ggsave()` or  `toHTML()` I described user can create its own html and manipulate it somehow (including in the notebook outputs).

alshan replied:
But `NotebookFrontendContext` already has `getDisplayHtml()`, which has a more precise name considering its purpose. What's the need for `getHtml()`?

alshan replied:
`to_html()` in Python is a convenience method. Same as `ggsave()` basically but also can return html as string in case `path` is not provided.

alshan replied:
Wait ) Just realized there is a bunch of `getHtml()` methods in `NotebookFrontendContext` already.

AndreiKingsley commented:
@alshan, we can fix the version of the lets-plot library (and, but not necessarily - js) when compiling the artifact with integration. This will help user to avoid version incompatibility. That is, there will be only one version in the descriptor - Lets-Plot Kotlin, one dependency that will load the lpk of this version and lp corresponding to it (bundled at compilation). What do you think?

AndreiKingsley commented:
And of course it will avoid cases when the integration code is written on a version incompatible with the specified in argument.

alshan commented:
Do you mean drop arguments in `%use lets-plot` ?

AndreiKingsley commented:
Yes. There will be only one dependency on lets-plot-kotlin-jupyter, that will load lets-plot-kotlin with the same version and lets-plot artifacts with tge corresonding for this lpk versions.

alshan commented:
Sounds good.

AndreiKingsley commented:
Extracted integrations to separated modules `jupyter` and `geotools-jupyter`. These modules creates artifacts with all LP and LPK implementation (and all geotools implementations for `geotools-jupyter`) and can be used in notebooks with updated descriptors ([this PR with descriptors](https://github.com/Kotlin/kotlin-jupyter-libraries/pull/421.) should be merged with the versions of integration artifacts inserted).
Also added "util" module and artifact - it contains Lets-Plot spec serialization and deserialization methods. It's used internally in Integration and should be used externally in Kotlin Notebook plugin (now it uses "kandy-util" similar artifact, but it's better to store it here, since it's related to Lets-Plot, for future changes).

alshan commented on toolkit/util/build.gradle.kts:
```diff
@@ -0,0 +1,66 @@
+plugins {
```
We like self-descriptive code. Could we rename "utils" to something more specific? Who's the utils client?

AndreiKingsley replied:
Well, this module is responsible for spec serialization. Also, I'd like to move here a specific logic, that is used  for example in Kotlin Notebook - applying the flavour for spec if it is not set (this is necessary for auto-applying the environment theme). 
So generally, this is useful stuff for those who create a Lets-plot spec on the client and send it (like Kotlin Notebook).

alshan replied:
The package is `package org.jetbrains.letsPlot.toolkit.jupyter`.

If this project is jupyter specific I would propose to rename it to "jupyter-util".

On the other hand, if it's not jupyter specific then we better remove "jupyter" from the package name and rename the project to smth like "kotlinx-serialization" (and the package - accordingly).

AndreiKingsley replied:
Ok, may be just "serialization"?

alshan replied:
I'm not sure now. This is generic serialization code not related to LP or Jupyter. Strictly speaking it shouldn't be here at all. Do you use it in Jupyter integration?

AndreiKingsley replied:
Yes, integration uses serialization of plot spec to pass it to Notebook. Notebook uses deserialisation part of this code to get the spec (and then use it in for Swing rendering).

AndreiKingsley replied:
As I said above, this can be very useful for users who want to transfer plot spec (create it on the client and send it to the server, for example).

alshan replied:
Or, let's remove mentioning of letsPlot and spec:
```

typealias LetsPlotSpec = Map<String, Any>

fun serializeSpec(spec: LetsPlotSpec): JsonElement {
    return serialize(spec)
}
```

Next, this module is looking as a patch/missing feature for kotlinx.serilization.json wherefore we can name the package as "org.jetbrains.letsPlot.toolkit.kotlinx.json"  and the project name would be "kotlinx-json".
What do you think?

AndreiKingsley replied:
But why would we remove the mention of lets-plot? We can more generally say that it is a `serialisation of a Map that contains nested only Map, List and primitives`, but it is exactly a serialisation of Lets-Plot spec. Yes, the module does not use Lets-plot directly, but this function is restricted to work specifically with Lets-Plot spec. If the spec structure is changed in the future (like adding a new possible type, or something like that), then we will also have to change this logic. So this implementation depends directly on Lets-Plot.

alshan replied:
But JSON is essentially just Maps, Lists and primitives. I don't see anything that is specific for lets-plot.

AndreiKingsley replied:
Ok

AndreiKingsley replied:
Renamed to `json`. All Lets-plot mentions are removed.

alshan commented on toolkit/util/src/main/kotlin/org/jetbrains/letsPlot/toolkit/jupyter/util/serializeSpec.kt:
```diff
@@ -0,0 +1,84 @@
+    }
+}
+
+fun deserializeSpec(json: JsonElement): LetsPlotSpec {
```
Can be replaced with `deserializeMap` with an additional parameter controlling handling on `null` (i.e. keep/drop/error).

ileasile commented on toolkit/util/src/main/kotlin/org/jetbrains/letsPlot/toolkit/jupyter/util/serializeSpec.kt:
```diff
@@ -0,0 +1,84 @@
+        json is JsonNull -> null
+        json.isString -> json.content
+        else -> {
+            json.booleanOrNull ?: json.doubleOrNull ?: error("Unknown JSON primitive type: [$json]")
```
@alshan we don't try to parse int here (and most likely it will result in an error) so I would say it's not completely universal

alshan replied:
AFAIK there is only one numeric type in JSON - double-precision floating-point and decimal point in the string representation is just an option. Accordingly, the code above looks right and universal.
